transaction/core/tests/input_rules.rs

```diff
@@ -63,25 +72,486 @@ fn test_input_rules_verify_max_tombstone() {
+    get_first_rules_mut(&mut tx).partial_fill_change = Some(change_committed_amount(
+        &revealed_tx_outs[0],
+        Amount::new(2000, 0.into()),
+    ));
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::FractionalOutputAmountDoesNotRespectFillFraction)
+    );
+}
+
+// Test that input rules verification is working for Tx with a mix of required
+// and partial fill outputs
+#[test]
+fn test_input_rules_verify_mixed_required_and_partial_fill_outputs() {
+    let block_version = BlockVersion::THREE;
+    let (mut tx, revealed_tx_outs) = get_input_rules_test_tx(block_version);
+
+    // Set the partial fill change output to be 1200, so fill fraction is 1/6
+    get_first_rules_mut(&mut tx).partial_fill_change = Some(change_committed_amount(
+        &revealed_tx_outs[0],
+        Amount::new(1200, 0.into()),
+    ));
+
+    // Add a partial fill output, by going to 6x the revealed tx out
+    get_first_rules_mut(&mut tx)
+        .partial_fill_outputs
+        .push(change_committed_amount(
+            &revealed_tx_outs[1],
+            Amount::new(6000, 0.into()),
+        ));
+
+    // Everything should be okay right now
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // Add a required output, using the last tx out
+    get_first_rules_mut(&mut tx)
+        .required_outputs
+        .push(revealed_tx_outs[3].tx_out.clone());
+
+    // Everything should be okay right now
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // Try changing the partial fill output by increasing the ask slightly, the Tx
+    // should be invalid
+    get_first_rules_mut(&mut tx).partial_fill_outputs[0] =
+        change_committed_amount(&revealed_tx_outs[1], Amount::new(6001, 0.into()));
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::FractionalOutputAmountDoesNotRespectFillFraction)
+    );
+
+    // Reduce partial fill change offer slightly, things should be okay now
+    get_first_rules_mut(&mut tx).partial_fill_change = Some(change_committed_amount(
+        &revealed_tx_outs[0],
+        Amount::new(1199, 0.into()),
+    ));
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // Things should still be okay on the ask up to 6025, but bad at 6026
+    get_first_rules_mut(&mut tx).partial_fill_outputs[0] =
+        change_committed_amount(&revealed_tx_outs[1], Amount::new(6026, 0.into()));
+    assert_matches!(
+        get_first_rules(&tx).verify(block_version, &tx),
+        Err(InputRuleError::FractionalOutputAmountDoesNotRespectFillFraction)
+    );
+    get_first_rules_mut(&mut tx).partial_fill_outputs[0] =
+        change_committed_amount(&revealed_tx_outs[1], Amount::new(6025, 0.into()));
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // Try tweaking the required output
+    get_first_rules_mut(&mut tx).required_outputs[0] =
```

Comment:
Am I correct in thinking that the reason this fails is because even though there is an output of 1000, the required output requires an output of EXACTLY 987 and thus 1000 doesn't match?

As a corollary, if there is a partial fill ratio you want that would result in giving 1001 to originator, that would actually result in you needing to give 1000 + 1001 to originator in order to match the required output, but if you could get the partial fill ratio to exactly 1000, you only give them 1000 and the sci would be satisfied?